### PR TITLE
Cpp refactor 2

### PR DESCRIPTION
**Pull request by BeckmaR**
Modularized the CPP generator, using a completely new way to configure the generator via annotations.

### CODE REVIEW

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend [hunk 1]
```diff
@@ -76,14 +76,6 @@ class Naming {
 		return entry.moduleName.toFirstUpper
 	}
 
-	def filterNullOrEmptyAndJoin(Iterable<CharSequence> it) {
-		filter[!it?.toString.nullOrEmpty].join('\n')
-	}
-
-	def filterNullOrEmptyAndJoin(List<String> it) {
-		filter[!it?.toString.nullOrEmpty].join('\n')
-	}
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend]: The first function was used once, the second one never, so I inlined and deleted them.

Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend]: +1

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CodeGeneratorFragmentProvider.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (c) 2018 committers of YAKINDU and others.
+ * All rights reserved. This program and the accompanying materials
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CodeGeneratorFragmentProvider.java]: 2019 ...

Comment (BeckmaR) [reply plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/CodeGeneratorFragmentProvider.java]: Whoopsie

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/providers/StatemachineClassDeclaration.xtend [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	@Inject protected extension InterfaceFunctions
+	@Inject protected extension TimingFunctions
+	@Inject protected extension TracingFunctions
+	@Inject protected extension LifecycleFunctions
+	
+	override get(ExecutionFlow it, IGenArtifactConfigurations artifactConfigs) {
+		generateClass(it, artifactConfigs).generate
+	}
+	
+	def generateClass(ExecutionFlow it, extension IGenArtifactConfigurations artifactConfigs) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/providers/StatemachineClassDeclaration.xtend]: generateClass is not a name that really matches the intend. 'generate' should be used for text transformations. Here the class definition is assembled or build.

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	protected CharSequence name
+	protected CharSequence comment
+	
+	protected List<CharSequence> superTypes = newArrayList
+	
+	protected List<CharSequence> publicMembers = newArrayList
+	protected List<CharSequence> protectedMembers = newArrayList
+	protected List<CharSequence> privateMembers = newArrayList
+	
+	def ClassDeclaration name(CharSequence name) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend]: setter should be named 'setName'. This is cleaner from a Java/Xtend perspective and better supported by Xtend. So initialized of ClassDeclaration instances can be written more nicely.

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		this.name = name
+		this
+	}
+	
+	def ClassDeclaration comment(CharSequence comment) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend]: same here ...

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		this.comment = comment
+		this
+	}
+	
+	def ClassDeclaration superType(CharSequence superType) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend]: same ...

File: plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		superTypes.add(superType)
+		this
+	}
+	
+	def ClassDeclaration publicMember(CharSequence member) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.cpp/src/org/yakindu/sct/generator/cpp/templates/ClassDeclaration.xtend]: same ...

**Pull request merged**